### ISSUE TITLE
Home - Upcoming events spinner keeps running

### ISSUE THREAD

**Issue by hkist** [jhuckaby/Cronicle#75]
## Summary

At the page Home - Upcoming events spinner keeps running

## Steps to reproduce the problem

Scheduled events (currently 6 till 900).

## Your Setup

Single master with multiple slaves.

### Operating system and version?

Ubuntu 18.04

### Node.js version?

8.10.0

### Cronicle software version?

8.0.11 (same problem with 8.0.10)

### Are you using the filesystem as back-end storage, or S3/Couchbase?

filesystem

### Can you reproduce the crash consistently?

Same result with different browsers en machines.

**Comment by jhuckaby** [comment]
Can you explain what you mean by this:

> Scheduled events (currently 6 till 900).

I don't know what "6 till 900" means.

It would be really helpful if I could get a look at your schedule data itself.  As long as you don't have any sensitive information in there, here is a command to export it to JSON on the command-line.  You can copy/paste it here, or e-mail it to me privately: jhuckaby AT gmail.

```
/opt/cronicle/bin/storage-cli.js list_get global/schedule
```

Thanks!

**Comment by jhuckaby** [comment]
I never received an e-mail from you, but I did release [Cronicle v0.8.12](https://github.com/jhuckaby/Cronicle/releases/tag/v0.8.12) with a potential bug fix on the Home screen.  I found one case where the upcoming events code might get stuck when simulating events to build the list.

Also, I should note that the "Upcoming Events" feature on the Home tab is all running inside a [web worker](https://developer.mozilla.org/en-US/docs/Web/API/Web_Workers_API/Using_web_workers), so this may be causing some issues for you if you have special security settings in your browser or on your corporate network.  Perhaps in a future version I'll move that code out of a web worker and back onto the main thread, if I can ensure it will never slow things down.

**Comment by jhuckaby** [comment]
Released [Cronicle v0.8.13](https://github.com/jhuckaby/Cronicle/releases/tag/v0.8.13) with another bug fix which may help this issue.  The maximum run time for calculating the upcoming events is now capped at 3 seconds.

**Comment by hkist** [comment]
Problem solved after upgrade to 0.8.12. However, sins then the upcoming eventlist is showing only one type of event (runs every minute). After switching to another tab (Completed), the rest of the events appears in the list. Upgrade to 0.8.14 didn't solve the problem. We have now deleted the every minute jobs.
I did send you the mail. Maybe in the spambox?
In general Cronicle is running fine!

**Comment by jhuckaby** [comment]
Hmm, well, I'm glad it is at least in a "working" state now.  I think your schedule may be just jam-packed with events, and the predictor code is choking on it.  What I did was add several pieces of code to short circuit and abort the loop if it ran for more than 3 seconds, or added more than 1,000 upcoming events.  Due to these changes, it may not be able to show all your upcoming events.

I did check my spam folder but couldn't find anything.  It's okay tho, I think we have an idea what is happening.  Honestly, there isn't a lot I can do, as this is a scalability issue (see issue #71 ).  The schedule is only designed for a handful of items, not 900.

Anyway, thank you for this issue report.  I will keep it in mind when I readdress scalability for v2.0.
